### DISCUSSION TITLE
How to separate tests for web and native?

### DISCUSSION THREAD

**Discussion by victorfernandesraton** [necolas/react-native-web#1967] [Q&A] [answered]
I need separate tests for web and native in a multiplatform project, I just founding something like this:

- Run web test only
```bash
npm run testWeb
```
- Run native test only
```bash
npm run testNative
```
- Run both test
```bash
npm run test
```

**Comment by riteshoyo** [comment]
@victorfernandesraton Did you able to achieve this? Working on the similar lines and facing some challenges.

**Comment by victorfernandesraton** [comment]
> @victorfernandesraton Did you able to achieve this? Working on the similar lines and facing some challenges.

1y after them and same problem...

**Comment by necolas** [comment, marked as answer]
You can use jest [projects](https://jestjs.io/docs/configuration#projects-arraystring--projectconfig) to configure different test environments, and match against different file extensions for platform-specific tests. Configure jest to produce platform-specific snapshots too. Then when you run jest it can correctly run any combination of web-only, native-only, and multiplatform unit tests.

For example:

```js
// jest.config.js

module.exports = {
  projects: [
    {
      displayName: 'web',
      moduleNameMapper: {
        // alias react-native
        '^react-native$': 'react-native-web'
      },
      // store DOM snapshots
      snapshotResolver: '<rootDir>/configs/jest-web-resolver.js',
      // use jsdom
      testEnvironment: 'jsdom',
      testMatch: [
        // all multi-platform tests
        '<rootDir>/src/*/__tests__/*-test.js',
        // DOM-only tests
        '<rootDir>/src/*/__tests__/*-test.web.js'
      ]
    },
    {
      displayName: 'native',
      // store React Native snapshots
      snapshotResolver: '<rootDir>/configs/jest-native-resolver.js',
      // use Node.js
      testEnvironment: 'node',
      testMatch: [
        // all multi-platform tests
        '<rootDir>/packages/react-strict-dom/tests/*-test.js',
        // native-only tests
        '<rootDir>/packages/react-strict-dom/tests/*-test.native.js'
      ],
    }
  ]
}
```

If you want platform-specific test commands, you can run individual projects like this:

```
jest --selectProjects web
jest --selectProjects native
```

**Comment by CAMOBAP** [comment]
Have you also see `snapshot files obsolete` warnings?
